File: packages/server/src/auth/http-auth.ts
import type { FastifyReply, FastifyRequest } from "fastify"

export function parseCookies(header: string | undefined): Record<string, string> {
  const result: Record<string, string> = {}
  if (!header) return result

  const parts = header.split(";")
  for (const part of parts) {
    const index = part.indexOf("=")
    if (index < 0) continue
    const key = part.slice(0, index).trim()
    const value = part.slice(index + 1).trim()
    if (!key) continue
    result[key] = decodeURIComponent(value)
  }
  return result
}

export function isLoopbackAddress(remoteAddress: string | undefined): boolean {
  if (!remoteAddress) return false
  if (remoteAddress === "127.0.0.1" || remoteAddress === "::1") return true
  if (remoteAddress === "::ffff:127.0.0.1") return true
  return false
}

export function wantsHtml(request: FastifyRequest): boolean {
  const accept = (request.headers["accept"] ?? "").toString().toLowerCase()
  return accept.includes("text/html") || accept.includes("application/xhtml")
}

export function sendUnauthorized(request: FastifyRequest, reply: FastifyReply) {
  if (request.method === "GET" && !request.url.startsWith("/api/") && wantsHtml(request)) {
    reply.redirect("/login")
    return
  }

  reply.code(401).send({ error: "Unauthorized" })
}

File: packages/server-rust/src/auth/http_auth.rs
use std::collections::HashMap;

pub fn parse_cookies(header: Option<&str>) -> HashMap<String, String> {
    let mut result = HashMap::new();
    let header = match header {
        Some(h) => h,
        None => return result,
    };

    for part in header.split(';') {
        let trimmed = part.trim();
        if let Some(index) = trimmed.find('=') {
            let key = trimmed[..index].trim().to_string();
            let value = trimmed[index + 1..].trim().to_string();
            if !key.is_empty() {
                result.insert(key, urlencoding_decode(&value));
            }
        }
    }

    result
}

fn urlencoding_decode(value: &str) -> String {
    // Simple percent-decoding for cookies
    let mut result = String::with_capacity(value.len());
    let mut chars = value.chars();
    while let Some(c) = chars.next() {
        if c == '%' {
            let hex: String = chars.by_ref().take(2).collect();
            if hex.len() == 2 {
                if let Ok(byte) = u8::from_str_radix(&hex, 16) {
                    result.push(byte as char);
                    continue;
                }
            }
            result.push('%');
            result.push_str(&hex);
        } else {
            result.push(c);
        }
    }
    percent_encoding::percent_decode_str(value)
        .decode_utf8_lossy()
        .to_string()
}

pub fn is_loopback_address(remote_addr: Option<&str>) -> bool {
    match remote_addr {
        Some("127.0.0.1") | Some("::1") | Some("::ffff:127.0.0.1") => true,
        _ => false,
    }
}

pub fn wants_html(accept_header: Option<&str>) -> bool {
    let accept = accept_header.unwrap_or("").to_lowercase();
    accept.contains("text/html") || accept.contains("application/xhtml")
}

pub fn is_loopback_host(host: &str) -> bool {
    host == "127.0.0.1" || host == "::1" || host.starts_with("127.")
}
